tests/kernel/interrupt/src/interrupt_offload.c

```diff
@@ -0,0 +1,266 @@
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(thread_prio),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* start a work queue thread if not existing */
+	if (!wq_already_start) {
+		k_work_queue_start(&wq_queue, wq_stack, STACK_SIZE,
+		1, NULL);
+
+		wq_already_start = true;
+	}
+
+	/* wait for thread start */
+	k_sem_take(&start_sem, K_FOREVER);
+
+	for (int i = 0; i < NUM_WORK; i++) {
+
+		switch (case_type) {
+		case TEST_OFFLOAD_MULTI_JOBS:
+			trigger_offload_interrupt(real_irq,
+					(void *)&offload_work[i]);
+		break;
+		case TEST_OFFLOAD_INDETICAL_JOBS:
+			trigger_offload_interrupt(real_irq,
+					(void *)&offload_work[0]);
+
+			/*
+			 * An 1 us delay for offload the same job is
+			 * needed, or it will cause an infinite loop.
+			 */
+			k_usleep(1);
+		break;
+		default:
+			ztest_test_fail();
+		}
+	}
+
+	/* wait for all offload job complete */
+	k_sem_take(&end_sem, K_FOREVER);
+
+	k_usleep(1);
+
+	zassert_equal(orig_t_keep_run, 1,
+			"offload job done, the original thread run");
+
+	k_thread_abort(tid);
+}
+
+/**
+ * @brief Test interrupt offload work to multiple jobs
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details Validate isr can offload workload to multi work queue, and:
+ *
+ * - If the priority of the original thread < offload job, offload jobs
+ *   could execute immediately.
+ *
+ * - If the priority of the original thread > offload job, offload jobs
+ *   will not execute immediately.
+ *
+ * We test this by irq_offload().
+ */
+void test_isr_offload_job_multiple(void)
+{
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_MULTI_JOBS, false);
+
+	offload_job_prio_higher = false;
+	run_test_offload(TEST_OFFLOAD_MULTI_JOBS, false);
+}
+
+/**
+ * @brief Test interrupt offload work to identical jobs
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details Validate isr can offload workload to work queue, and all
+ * the offload jobs use the same thread entry, and:
+ *
+ * - If the priority of the original thread < offload job, offload jobs
+ *   could execute immediately.
+ *
+ * - If the priority of the original thread > offload job, offload jobs
+ *   will not execute immediately.
+ *
+ * We test this by irq_offload().
+ */
+void test_isr_offload_job_identi(void)
+{
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_INDETICAL_JOBS, false);
+
+	offload_job_prio_higher = false;
+	run_test_offload(TEST_OFFLOAD_INDETICAL_JOBS, false);
+}
+
+/**
+ * @brief Test interrupt offload work by dynamic interrupt
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details Validate isr can offload workload to work queue, and the
+ * offload jobs could execute immediately base on it's priority.
+ * We test this by dynamic interrupt.
+ */
+void test_isr_offload_job(void)
+{
+	if (!IS_ENABLED(CONFIG_DYNAMIC_INTERRUPTS)) {
+		ztest_test_skip();
+	}
+
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_MULTI_JOBS, true);
+	run_test_offload(TEST_OFFLOAD_INDETICAL_JOBS, false);
```

Comment:
IDENTICAL. Use search to find all misprints in that word.

Reply:
Thanks for correcting this. Got it fixed.